samples: mesh: nrf52: revised implementation, fixed bugs, upgrade further to pass PTS test, upgrade coding

vikrant8052 opened the pull request:
Removed redundancy from App & revised overall implementation.
Fixed Bugs. Upgraded coding style. Added & replaced comments to
improve code readability.

Upgrade State Binding algorithm by adding ceiling() & constrain
functions.

If particular Server model receive Prohibited values as per
Mesh Model Specification then that message will not get entertained.

Now proper status code will get send if user upgrade
light_lightness_range_state & light_ctl_temperature_range_state.

Added time stamp based message filtering as per Bluetooth Mesh
Developer study guide for each Server. Now even if message
source address & TID is same even after 6 seconds then
that will get proceed otherwise will get ignored.

Added transition time support. With this it was possible to implement gen_move_set/gen_move_set_unack message handlers.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
Please do not merge. It is basically created to resolve bugs which are not yet solved as per PTS test.

vikrant8051 commented:
**MMDL/SR/GOO/BV-02-C FAIL** https://hastebin.com/futahusiru.php

        It failed since transition state concept is not implemented.

vikrant8051 commented:
MMDL/SR/GOO/BV-02-C
FAIL https://hastebin.com/futahusiru.php https://hastebin.com/siwivonafe.css	
>> because of absence of Transition Time based implementation.

MMDL/SR/GLV/BV-02-C
FAIL https://hastebin.com/horijapaqi.php https://hastebin.com/uluvuceqag.css	
>> because of absence of Transition Time based implementation.

MMDL/SR/GLV/BV-03-C
FAIL https://hastebin.com/ohanupezot.sql https://hastebin.com/tubipuvequ.css
>> because of absence of Transition Time based implementation.

MMDL/SR/GLV/BV-05-C
FAIL https://hastebin.com/lovuyurana.php https://hastebin.com/evawefahik.css
>> here I think, last test's TID == current test TID. Hence it get skipped.

MMDL/SR/GPOO/BV-02-C
FAIL https://hastebin.com/aseburahir.php https://hastebin.com/muxapovuqu.css	
>> because of absence of saving OnPowerUp state on SoC flash.

MMDL/SR/GPOO/BV-03-C
FAIL https://hastebin.com/fohejofoxu.php https://hastebin.com/goxigaroxe.css	
>> because of absence of Transition Time based implementation.

MMDL/SR/GPOO/BV-04-C
FAIL https://hastebin.com/ubejuxubem.php https://hastebin.com/punogofota.css	
>> here I think, last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-02-C
FAIL https://hastebin.com/ozayisesud.php https://hastebin.com/owuyixabuj.css	
>> because of failure in proper rounding off float variable

MMDL/SR/LLN/BV-06-C
FAIL https://hastebin.com/etogorotaz.php https://hastebin.com/awiwoxofaj.css	
>> here I think, last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-07-C
FAIL https://hastebin.com/uyinuwolif.php https://hastebin.com/ezukubatuh.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-08-C
FAIL https://hastebin.com/iqugojukok.php https://hastebin.com/etafabemab.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-09-C
FAIL https://hastebin.com/beceteqapi.php https://hastebin.com/dajinuqehi.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-10-C
FAIL https://hastebin.com/osegacazuh.php https://hastebin.com/ihawugipol.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-11-C
FAIL https://hastebin.com/vinolobive.php https://hastebin.com/utitobokes.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-12-C
FAIL https://hastebin.com/ihinefasup.php https://hastebin.com/cuqasavaxa.css	
>> TID is common for Lightness Actual state & Light Lightness Linear state.
>> So here too beacuse of last_TID == current TID, test failed.

MMDL/SR/LLN/BV-13-C
FAIL https://hastebin.com/uyacuyayug.php https://hastebin.com/ubadejarar.css	
>> TID is common for Lightness Actual state & Light Lightness Linear state.
>> So here too beacuse of last_TID == current TID, test failed.

MMDL/SR/LLN/BV-14-C
FAIL https://hastebin.com/acuxeseham.php https://hastebin.com/jaxezaveni.css	
>> TID is yet = 0, hence failed.

MMDL/SR/LLN/BV-15-C
FAIL https://hastebin.com/rorohelucu.php https://hastebin.com/dupamuzelo.css	
>> Here TID set to 255....hence things proceed further. But failed
>> because of absence of Transition Time based implementation.

MMDL/SR/LLN/BV-16-C
FAIL https://hastebin.com/piqacakaru.php https://hastebin.com/ubuxasuraf.css	
>> because of failure in proper rounding off float variable

MMDL/SR/LLN/BV-17-C
FAIL https://hastebin.com/ijitaxamoh.php https://hastebin.com/qujojexure.css
>> because of absence of saving OnPowerUp state on SoC flash.

MMDL/SR/LLNS/BI-01-C
FAIL https://hastebin.com/aliruzagel.php https://hastebin.com/juqejiroka.css
https://github.com/zephyrproject-rtos/zephyr/pull/8926													
>> This may solve the issue. It would send 0x01 instead 0x02 as status code.

MMDL/SR/LCTL/BV-05-C
FAIL https://hastebin.com/orixiviyij.m https://hastebin.com/atofawilec.css
>> because of absence of Transition Time based implementation.

MMDL/SR/LCTL/BV-06-C
FAIL https://hastebin.com/owikejifes.m https://hastebin.com/ewuqidahow.css
>> here I think, last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LCTL/BI-01-C
FAIL https://hastebin.com/ecugijezor.m https://hastebin.com/ohejazulal.css
>> issue has solved in PR:8793

MMDL/SR/LCTLS/BI-01-C
FAIL https://hastebin.com/cawisevafo.m https://hastebin.com/umoyepukem.css
>> issue will get solved with PR:8926

MMDL/SR/LCTLS/BI-02-C
FAIL https://hastebin.com/ocahipadir.m https://hastebin.com/ibusomupad.css
>> issue has solved in PR:8793

vikrant8051 commented:
@jhedberg @MariuszSkamra 
FYI, This is summary of previous PTS logs with reason why we get errors. Still working to make things better.....


MMDL/SR/GOO/BV-02-C
FAIL https://hastebin.com/futahusiru.php https://hastebin.com/siwivonafe.css	
>> because of absence of Transition Time based implementation.

MMDL/SR/GLV/BV-02-C
FAIL https://hastebin.com/horijapaqi.php https://hastebin.com/uluvuceqag.css	
>> because of absence of Transition Time based implementation.

MMDL/SR/GLV/BV-03-C
FAIL https://hastebin.com/ohanupezot.sql https://hastebin.com/tubipuvequ.css
>> because of absence of Transition Time based implementation.

MMDL/SR/GLV/BV-05-C
FAIL https://hastebin.com/lovuyurana.php https://hastebin.com/evawefahik.css
>> here I think, last test's TID == current test TID. Hence it get skipped.

MMDL/SR/GPOO/BV-02-C
FAIL https://hastebin.com/aseburahir.php https://hastebin.com/muxapovuqu.css	
>> because of absence of saving OnPowerUp state on SoC flash.

MMDL/SR/GPOO/BV-03-C
FAIL https://hastebin.com/fohejofoxu.php https://hastebin.com/goxigaroxe.css	
>> because of absence of Transition Time based implementation.

MMDL/SR/GPOO/BV-04-C
FAIL https://hastebin.com/ubejuxubem.php https://hastebin.com/punogofota.css	
>> here I think, last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-02-C
FAIL https://hastebin.com/ozayisesud.php https://hastebin.com/owuyixabuj.css	
>> because of failure in proper rounding off float variable

MMDL/SR/LLN/BV-06-C
FAIL https://hastebin.com/etogorotaz.php https://hastebin.com/awiwoxofaj.css	
>> here I think, last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-07-C
FAIL https://hastebin.com/uyinuwolif.php https://hastebin.com/ezukubatuh.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-08-C
FAIL https://hastebin.com/iqugojukok.php https://hastebin.com/etafabemab.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-09-C
FAIL https://hastebin.com/beceteqapi.php https://hastebin.com/dajinuqehi.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-10-C
FAIL https://hastebin.com/osegacazuh.php https://hastebin.com/ihawugipol.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-11-C
FAIL https://hastebin.com/vinolobive.php https://hastebin.com/utitobokes.css	
>> last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LLN/BV-12-C
FAIL https://hastebin.com/ihinefasup.php https://hastebin.com/cuqasavaxa.css	
>> TID is common for Lightness Actual state & Light Lightness Linear state.
>> So here too beacuse of last_TID == current TID, test failed.

MMDL/SR/LLN/BV-13-C
FAIL https://hastebin.com/uyacuyayug.php https://hastebin.com/ubadejarar.css	
>> TID is common for Lightness Actual state & Light Lightness Linear state.
>> So here too beacuse of last_TID == current TID, test failed.

MMDL/SR/LLN/BV-14-C
FAIL https://hastebin.com/acuxeseham.php https://hastebin.com/jaxezaveni.css	
>> TID is yet = 0, hence failed.

MMDL/SR/LLN/BV-15-C
FAIL https://hastebin.com/rorohelucu.php https://hastebin.com/dupamuzelo.css	
>> Here TID set to 255....hence things proceed further. But failed
>> because of absence of Transition Time based implementation.

MMDL/SR/LLN/BV-16-C
FAIL https://hastebin.com/piqacakaru.php https://hastebin.com/ubuxasuraf.css	
>> because of failure in proper rounding off float variable

MMDL/SR/LLN/BV-17-C
FAIL https://hastebin.com/ijitaxamoh.php https://hastebin.com/qujojexure.css
>> because of absence of saving OnPowerUp state on SoC flash.

MMDL/SR/LLNS/BI-01-C
FAIL https://hastebin.com/aliruzagel.php https://hastebin.com/juqejiroka.css
https://github.com/zephyrproject-rtos/zephyr/pull/8926													
>> This may solve the issue. It would send 0x01 instead 0x02 as status code.

MMDL/SR/LCTL/BV-05-C
FAIL https://hastebin.com/orixiviyij.m https://hastebin.com/atofawilec.css
>> because of absence of Transition Time based implementation.

MMDL/SR/LCTL/BV-06-C
FAIL https://hastebin.com/owikejifes.m https://hastebin.com/ewuqidahow.css
>> here I think, last test's TID == current test TID. Hence it get skipped.

MMDL/SR/LCTL/BI-01-C
FAIL https://hastebin.com/ecugijezor.m https://hastebin.com/ohejazulal.css
>> issue has solved in PR:8793

MMDL/SR/LCTLS/BI-01-C
FAIL https://hastebin.com/cawisevafo.m https://hastebin.com/umoyepukem.css
>> issue will get solved with PR:8926

MMDL/SR/LCTLS/BI-02-C
FAIL https://hastebin.com/ocahipadir.m https://hastebin.com/ibusomupad.css
>> issue has solved in PR:8793

vikrant8051 commented:
MMDL/SR/LLNS/BI-01-C
https://hastebin.com/aliruzagel.php .... I think failed because received status code is 0x02 instead of 0x01.

MMDL/SR/LCTLS/BI-01-C
https://hastebin.com/cawisevafo.m .... I think failed because received status code is 0x02 instead of 0x01.

With this PR:8926 it would send 0x01. Kindly run PTS only for these two.

vikrant8051 commented:
@jhedberg @MariuszSkamra 
PR is ready to be but ...

Wrapping Light Lightness Actual state is very confusion.
Please refer 6.1.2.2.2 vs 3.3.2.2.3....don't know what to follow ? Need help.

vikrant8051 commented:
@MariuszSkamra 

MMDL/SR/LLNS/BI-01-C
https://hastebin.com/aliruzagel.php .... I think failed because received status code is 0x02 instead of 0x01.

MMDL/SR/LCTLS/BI-01-C
https://hastebin.com/cawisevafo.m .... I think failed because received status code is 0x02 instead of 0x01.

With this PR:8926 it would send 0x01. Kindly re-run PTS only for these two ( Please feed info to PTS that "persistent storage for STATE saving" & "transition time" support is absent)

vikrant8051 commented:
@jhedberg 

PR is ready from my side, so please review it.

MMDL/SR/LLN/BV-02-C -> solved
MMDL/SR/LLN/BV-16-C -> solved
MMDL/SR/LLNS/BI-01-C -> solved
MMDL/SR/LCTL/BI-01-C -> solved
MMDL/SR/LCTLS/BI-01-C -> solved
MMDL/SR/LCTLS/BI-02-C -> solved

Others failures are because of 
1) PTS own issue which is starting every new test with same TID value that is zero. Because of that Servers are not reacting to it. 
2) absence of Server's state saving on SoC flash
3) absence of transition time based implementation which is optional

vikrant8051 commented:
@jhedberg 

Please keep things on Hold.
FYI,
Now I've received response from SIG forum. I will inform you once clears my doubts.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c:
```diff
@@ -128,7 +128,7 @@ void update_light_state(void)
 void main(void)
 {
-	int err;
+	s32_t err;
```
bt_enable() returns int and not s32_t. Why did you change this?

vikrant8051 replied:
Ok I will make it int again

vikrant8051 replied:
done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/ble_mesh.c:
```diff
@@ -10,7 +10,7 @@
 #ifdef OOB_AUTH_ENABLE
 
-static int output_number(bt_mesh_output_action_t action, uint32_t number)
+static int output_number(bt_mesh_output_action_t action, u32_t number)
```
Wasn't this already fixed upstream? Could you try rebasing your PR on latest upstream master?

vikrant8051 replied:
I will keep things as it is.

vikrant8051 replied:
No, it wasn't get fixed in upstream

jhedberg replied:
I was looking at commit 4208642a536235477019b802f313b9802d16f2c9

But seems you're right: this commit missed the place you're fixing

vikrant8051 replied:
Done !!

vikrant8051 commented:
@jhedberg 
I've done something wrong while rebasing. :

vikrant8052 committed:
samples: mesh: nrf52: revised implementation

Removed redudandancy from App & revised overall implementation.
Fixed Bugs. Added & replaced comments to improve code readability.

Upgrade State Binding algorithm by adding ceiling() & constrain
functions.

If particular Server model receive Prohibited values as per
Mesh Model Specification then that message will not get entertained.

Now proper status code will get send if user upgrade
light_lightness_range_state & light_ctl_temperature_range_state.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg 
Mr. Martin from SIG helped me to clear my doubts. As per that & PTS requirements I've restructure things. Please consider it & review further.

Thank You !!

vikrant8051 commented:
@jhedberg Kindly wait. During testing I've found some bugs. I will repair them shortly.

vikrant8051 commented:
@jhedberg 
I didn't found any coding bug as of now. I changed PDK board (as client) &  provisioned everything using #meshctl instead of #nRFMesh. Now didn't found that previous bug while re-testing.

I think it was h/w button issue :(

vikrant8051 commented:
New implementation is tested & now ready to get merged.

vikrant8051 commented:
@jhedberg New implementation is tested & now ready to get merged.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -107,32 +107,92 @@ static float sqrt(float square)
+	is_value_within_range = false;
+
+	if (var > 0 && var <
+		light_lightness_srv_user_data.lightness_range_min) {
```
This is incorrectly aligned which makes it very confusing to look at, especially since the var assignment starts with the same alignment level. The split should happen after `&&` and the line should begin right after the `(` on the first line.

vikrant8051 replied:
done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -107,32 +107,92 @@ static float sqrt(float square)
+	if (var > 0 && var <
+		light_lightness_srv_user_data.lightness_range_min) {
+		var =
+		light_lightness_srv_user_data.lightness_range_min;
```
I'd suggest looking for shorter naming of your variables, if you have to do this kind of line splitting for only two tabs deep indentation.

vikrant8051 replied:
If I shorter the name that will reduce readability. I will go with your first suggestion

vikrant8051 replied:
Ok....going to modifiy **struct light_lightness_state** as ...

struct light_lightness_state {
	u16_t linear;
	u16_t actual;
	u16_t last;
	u16_t def;

	u8_t status_code;
	u16_t light_range_min;
	u16_t light_range_max;

	u8_t last_tid;
	u16_t last_tx_addr;
};

vikrant8051 replied:
done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -107,32 +107,92 @@ static float sqrt(float square)
+	} else {
+		is_value_within_range = true;
+	}
+	light_lightness_srv_user_data.actual = var;
```
Empty line always after }, unless the next line is also a }

vikrant8051 replied:
done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -149,232 +210,116 @@ static void state_binding(u8_t lightness, u8_t temperature)
 		tmp16 = gen_level_srv_root_user_data.level + 32768;
 		if (tmp16 > 0 && tmp16 <
-		    light_lightness_srv_user_data.lightness_range_min) {
+			light_lightness_srv_user_data.lightness_range_min) {
```
Looks like you've broken the alignment here and split the line in the wrong place. Split it after `&&` and align it with the ( like it used to be.

vikrant8051 replied:
done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -149,232 +210,116 @@ static void state_binding(u8_t lightness, u8_t temperature)
 			}
 		} else if (tmp16 >
-			   light_lightness_srv_user_data.lightness_range_max) {
+			light_lightness_srv_user_data.lightness_range_max) {
```
Same here. Why did you even change this line?

vikrant8051 replied:
done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -149,232 +210,116 @@ static void state_binding(u8_t lightness, u8_t temperature)
 		tmp16 = (u16_t) 65535 *
 			sqrt(((float) light_lightness_srv_user_data.linear /
-			      65535));
+			65535));
```
Same here.

vikrant8051 replied:
done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -149,232 +210,116 @@ static void state_binding(u8_t lightness, u8_t temperature)
-		light_ctl_srv_user_data.lightness =
-			light_lightness_srv_user_data.actual;
+		constrain_lightness_actual_state(
+			light_ctl_srv_user_data.lightness);
```
This looks quite messed up as well. I'd suggest rethinking your function and variable naming, since they're super long right now.

vikrant8051 replied:
Going to rename as ......constrain_light_actual_state

vikrant8051 replied:
done !!

jhedberg reviewed (dismissed):
Mostly coding style issues (some of which hurt readability quite badly, so it's important to fix them). I've yet to see a PTS test report from @MariuszSkamra - that's something that's also needed before merging.

vikrant8051 commented:
@jhedberg I've make all suggested changes.

jhedberg commented:
@vikrant8051 thanks. I'll still wait for the PTS results. In the future, could you please split your changes so that each commit contains a single logically independent change. That makes understanding the change history easier and also helps a lot with bisecting (git bisect) when bugs are found.

vikrant8051 commented:
@jhedberg @MariuszSkamra 
As per above mentioned reasons for failing PTS tests, one of the prime reason is PTS's dongle in some **consecutive** tests, send same TID value to particular Server. 

And as per #onoff_level_lighting_vnd_app, if any Server receives same TID value from same Source address, it assumes message has already proceed & current one is retransmitted one & hence Server ignores it.

vikrant8051 commented:
@jhedberg 
Hi, now going to add time stamp based message filter as per Bluetooth Mesh Developer study guide for each Server. So keep things on hold.

vikrant8052 committed:
samples: mesh: nrf52: added time stamp based message filtering

Added time stamp based message filtering as per Bluetooth Mesh
Developer study guide for each Server. Now even if message
source address & TID is same even after 6 seconds then
that will get proceed otherwise will get ignored.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg Please go through with latest commit for review.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -584,6 +584,57 @@ static void gen_move_set_unack(struct bt_mesh_model *model,
+
+	tmp16 = state->level + delta;
+	if (tmp16 < -32768) {
+		tmp16 = -32768;
```
Is this really needed? Isn't -32768 the minimum possible value a 16-bit signed integer can hold, i.e. INT16_MIN?

vikrant8051 replied:
done !!

vikrant8051 replied:
As of now I've removed content under gen_move_set/gen_move_set_unack. Since for that transition time based support is necessary.

vikrant8051 replied:
Revised content under ge n_move_set/ge n_move_set_unack. So mentioned issue has resolved !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -584,6 +584,57 @@ static void gen_move_set_unack(struct bt_mesh_model *model,
+	if (tmp16 < -32768) {
+		tmp16 = -32768;
+	} else if (tmp16 > 32767) {
+		tmp16 = 32767;
```
This looks also unnecessary as 32767 is the maximum possible value a 16-bit integer can hold (INT16_MAX)

vikrant8051 replied:
done!!

vikrant8051 replied:
Revised content under gen_move_set/gen_move_set_unack. So mentioned issue has resolved !!

vikrant8051 replied:
done !!

vikrant8051 commented:
@jhedberg 
Is it mandatory for Servers to entertain transition time & delay fields in message ?

vikrant8051 commented:
@jhedberg 

```
now = k_uptime_get();
if (state->last_tid == tid && state->last_tx_addr == ctx->addr && (now - state->last_msg_timestamp <= 6000)) {
 	return true;
}
```

Is it really correct implementation ? what if previous sampling of timer happen before timer roll over & next one happen after that ? In that case, **now** could have some negative value. Am I right ?

MariuszSkamra reviewed (commented):
I've checked the tests that should pass with thi PR according to your comment https://github.com/zephyrproject-rtos/zephyr/pull/8926#issuecomment-405818929

MMDL/SR/LLN/BV-02-C -PASS
MMDL/SR/LLN/BV-16-C FAIL https://hastebin.com/numaxulibo.php
MMDL/SR/LLNS/BI-01-C FAIL https://hastebin.com/mevidiheki.php
MMDL/SR/LCTL/BI-01-C PASS
MMDL/SR/LCTLS/BI-01-C FAIL https://hastebin.com/eyipivisid.m
MMDL/SR/LCTLS/BI-02-C PASS

vikrant8052 committed:
samples: mesh: nrf52: added transition support

This commit adds transition time support. With this
it was possible to implement gen_move_set/gen_move_set_unack
message handlers.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg 
I've added support to entertain messages with transition time & delay. With that it was easy to implement gen_move_set/gen_move_set_unack message handlers.

I need your special attention/guidance for latest commit. Please review it throughly. 

@MariuszSkamra 
Thanks for latest PTS log. Happy to know that now there are only few failed log. I will repair things shorty.

Thank You to both of you !!

vikrant8052 committed:
samples: mesh: nrf52: range status code set to zero

Range status code set to Zero after sending status response.
This is as per PTS test requirements.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/transition.c:
```diff
@@ -0,0 +1,456 @@
+void temperature_transition_timer_handler(struct k_timer *dummy)
+{
+	s32_t tmp_s32;
+	u32_t tmp_u32;
```
Variable declarations should be in the smallest possible scope, so please move them there. Also rename them more sensibly. It seems like `new_level` and `new_temp` would be much more descriptive.

vikrant8051 replied:
I will rename those. But if moved it to inner scope then unnecessarily I've too declare them twice.. will it do ?

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/transition.c:
```diff
@@ -0,0 +1,456 @@
+void lightness_transition_timer_handler(struct k_timer *dummy)
+{
+	s32_t tmp_s32;
+	u32_t tmp_u32;
```
Move these to the smaller scopes where they're use and rename them more descriptively.

vikrant8051 replied:
Same as per above

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/transition.c:
```diff
@@ -0,0 +1,456 @@
+
+	state->tt_delta_uv_delta =
+		((float) abs(state->delta_uv - state->target_delta_uv)
+		 / state->tt_counter);
```
The operand should be on the upper line when splitting lines (/ in this case)

There are many other places like this - I assume you'll fix them without me having to comment on every single one.

vikrant8051 replied:
Ok.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c:
```diff
@@ -0,0 +1,206 @@
+{
+	bool bin;
+	u16_t tmp16;
+	float tmp;
```
Please stop using all these tmp variables and name your variables descriptively. It's really hard to follow the code otherwise. Should this have been `temp` instead of `tmp`? Same thing with tmp16.

vikrant8051 replied:
Why temp ?? ...it has used at multiple location.
It is there from beginning onwards. So why this sudden requirement pops up ?

jhedberg replied:
> It is there from beginning onwards. So why this sudden requirement pops up ?

The answer is really quite simple: I'm not a machine.

Your app is not small, and it just keeps growing. I have limited time to go through it and will for sure keep missing out on things in the future as well.

vikrant8051 replied:
I'm completely agree with you. But currently if possible ignore it. I will do such make-up related things via separate PR once we pass test PTS. 

And thanks for considering this App is not a small app.... :)

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c:
```diff
@@ -0,0 +1,206 @@
+
+void state_binding(u8_t lightness, u8_t temperature)
+{
+	bool bin;
```
What's `bin` short for?

vikrant8051 replied:
binary !!!!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c:
```diff
@@ -0,0 +1,206 @@
+
+#define MINDIFF 2.25e-308
+
+static float sqrt(float square)
```
Is there really no existing helper for this in our libc, etc?

vikrant8051 replied:
For that I've to add math.h which could increase ROM size. Plus we've to enable some features via make menuconfig. 

Again ... this is as it is from past few versions. Then why this sudden requirement pops-up ?

I will repairs such things in next PR. Currently my focus is to PASS PTS test requirements & make things better from algorithm/architecture point of view.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -1445,7 +1362,6 @@ static void light_ctl_temp_get(struct bt_mesh_model *model,
 	struct light_ctl_state *state = model->user_data;
 
 	bt_mesh_model_msg_init(msg, BT_MESH_MODEL_OP_2(0x82, 0x66));
-
```
This looks like a coding style fix. Those should be in a separate commit. Same for any other style fixes in this PR.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -1140,48 +1035,67 @@ static bool light_ctl_setunack(struct bt_mesh_model *model,
+	lightness_transition = LIGTHT_CTL_TT;
+
+	k_timer_start(&lightness_transition_timer, K_MSEC(5*delay),
+		      K_MSEC(100));
```
Where does the K_MSEC(100) come from? The model spec? If so, please put a code comment explaining this.

vikrant8051 replied:
I will repair other things which are mostly coding style issues.

But this is important one as of now. As per Mesh Model Specification 3.1.3.1, result
of calculation of transition time is always multiple of 100 milliseconds that is why I used K_MSEC(100). 

if TT = 0x85, then transition time = 10 Secs * 5 = 50 Secs = 50000 millisecs

So counter = 50000 / 100 = 50.

...& delta value which get added after every 100 milliseconds would be = delta/50

vikrant8051 replied:
Now 100 get replaced by Server's **cal_tt**.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -1140,48 +1035,67 @@ static bool light_ctl_setunack(struct bt_mesh_model *model,
 	}
 
-	if (temp < TEMP_MIN || temp > TEMP_MAX) {
+	switch (buf_len) {
```
I realize this buf_len stuff is probably coming from the SIGs Mesh guide, but you could avoid the extra variable by looking at buf->len instead, which will tell you how many bytes are left in the buffer (0 or 2 in your case). Same for the other places where you've used buf_len.

vikrant8051 replied:
done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -813,40 +664,58 @@ static void light_lightness_set_unack(struct bt_mesh_model *model,
-	if (state->last_tid == tid && state->last_tx_addr == ctx->addr) {
+	now = k_uptime_get();
+	if (state->last_tid == tid && state->last_tx_addr == ctx->addr &&
+	    (now - state->last_msg_timestamp <= 6000)) {
```
Perhaps K_SECONDS(6) instead of 6000? Same for the other similar place (which I wont comment on separately)

vikrant8051 replied:
It will add additional burden in calling a function.

jhedberg replied:
What makes you think that? K_SECONDS() is a macro that will evaluate to a build-time constant. Hint: macros are always capital case in Zephyr.

vikrant8051 replied:
I will do it shortly.

vikrant8051 replied:
I've to create separate commit since could not fixed commit 5 with commit 2 using rebase -i.

vikrant8051 commented:
Hi @jhedberg,

https://github.com/vikrant8051/test/tree/master/onoff_level_lighting_vnd_app

Here I've alloted different thread for each Server which internally calls own timer. Will it be better than current implementation ?

Please reply !! Thank You

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -626,6 +395,88 @@ static void gen_move_set_unack(struct bt_mesh_model *model,
+			lightness_transition = 0;
+			k_timer_stop(&lightness_transition_timer);
+			goto do_not_activate_timer;
+		}
```
Done !!

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -626,6 +395,88 @@ static void gen_move_set_unack(struct bt_mesh_model *model,
+			temperature_transition = 0;
+			k_timer_stop(&temperature_transition_timer);
+			goto do_not_activate_timer;
+		}
```
Done !!

vikrant8051 commented:
@jhedberg 
Going with separate thread for each Servers (who have to entertain TT & Delay fields).

vikrant8052 committed:
samples: mesh: nrf52: coding style, timestamp filtering under gen_move

Replaced 6000 with K_SECONDS(6) micro to increase code readability.
Added missing timestamp filtering in gen_move_set_unack message
handler.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg 
All checks have passed. I think most of the suggestions have applied. Please review.

Thanks for latest suggested improvements. 

But kindly not that I'm not able to provisioning node with latest Zephyr master tree.
It get provisioned if I used local git directory.

vikrant8051 commented:
@jhedberg Kindly don't review as of now since I'm upgrading the things.

vikrant8051 commented:
@jhedberg Now I think PR is ready for review.

vikrant8051 commented:
@MariuszSkamra @jhedberg 
If you are going to run PTS against this PR, then please activate transition time related things.

Note that status from Servers contains only Present state  & not Target state, Remaining Time....(since both parameter are optional for Server)

Hope this weekend it will get merge. Thank You !!

vikrant8052 committed:
samples: mesh: nrf52: improved coding style

Improved coding style. Removed blank lines.
Replaced -32767 & 32768 with INT16_MIN & INT16_MAX resp.
Corrected & improved comments, printk statements,
things in prj.conf.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

MariuszSkamra commented:
Hi,
I tested this PR against new PTS version 7.3. There might be some changes in tests scenario/behaviour. @vikrant8051 Please take a look. For me it looks like few of the tests that used to pass before fail now.

MMDL/SR-CL/COMP/BV-01-C FAIL https://hastebin.com/upeducodot.sql ( model 0x1007 (Generic Power OnOff Setup Server) requires model 0x1004 (Generic Default Transition Time Server), but it is not present.)
MMDL/SR/GOO/BV-01-C FAIL https://hastebin.com/cequkamoju.sql
MMDL/SR/GOO/BV-02-C FAIL https://hastebin.com/iboyurodaw.sql
MMDL/SR/GOO/BI-01-C PASS  
MMDL/SR/GLV/BV-01-C FAIL https://hastebin.com/remulijafo.sql
MMDL/SR/GLV/BV-02-C FAIL https://hastebin.com/riduvazara.sql
MMDL/SR/GLV/BV-03-C FAIL https://hastebin.com/nehofefogi.php
MMDL/SR/GLV/BV-04-C FAIL https://hastebin.com/igagiwoyus.php
MMDL/SR/GLV/BV-05-C FAIL https://hastebin.com/junudotope.sql
MMDL/SR/GLV/BV-06-C FAIL https://hastebin.com/zuwenimage.php
MMDL/SR/GPOO/BV-01-C PASS
MMDL/SR/GPOOS/BV-01-C PASS
MMDL/SR/GPOOS/BI-01-C PASS
MMDL/SR/LLN/BV-01-C FAIL https://hastebin.com/fuzudokoto.sql
MMDL/SR/LLN/BV-02-C FAIL https://hastebin.com/ewemesocaq.sql
MMDL/SR/LLN/BV-03-C PASS
MMDL/SR/LLN/BV-04-C PASS
MMDL/SR/LLN/BV-05-C PASS
MMDL/SR/LLN/BV-06-C FAIL https://hastebin.com/osuqadapuy.sql
MMDL/SR/LLN/BV-07-C FAIL https://hastebin.com/likiwirere.sql
MMDL/SR/LLN/BV-08-C FAIL https://hastebin.com/pocamimoxe.sql
MMDL/SR/LLN/BV-09-C FAIL https://hastebin.com/mohobogobi.sql
MMDL/SR/LLN/BV-10-C FAIL https://hastebin.com/adureratur.sql
MMDL/SR/LLN/BV-11-C FAIL https://hastebin.com/bapuholuvu.sql
MMDL/SR/LLN/BV-12-C FAIL https://hastebin.com/omesesahif.sql
MMDL/SR/LLN/BV-13-C FAIL https://hastebin.com/zilokekoyo.sql
MMDL/SR/LLN/BV-14-C FAIL https://hastebin.com/rujewutuza.php
MMDL/SR/LLN/BV-15-C FAIL https://hastebin.com/jipesuwiku.sql
MMDL/SR/LLN/BV-16-C FAIL https://hastebin.com/kusoqavire.sql
MMDL/SR/LLN/BV-17-C FAIL https://hastebin.com/ketagireyo.sql
MMDL/SR/LLNS/BV-01-C PASS 
MMDL/SR/LLNS/BV-02-C PASS
MMDL/SR/LLNS/BI-01-C FAIL https://hastebin.com/emubocaqaq.php
MMDL/SR/LCTL/BV-01-C FAIL https://hastebin.com/kaseyetise.sql
MMDL/SR/LCTL/BV-02-C PASS
MMDL/SR/LCTL/BV-03-C PASS
MMDL/SR/LCTL/BV-04-C FAIL https://hastebin.com/karubugeke.sql
MMDL/SR/LCTL/BV-05-C FAIL https://hastebin.com/gogaluxata.sql
MMDL/SR/LCTL/BV-06-C  FAIL https://hastebin.com/ubanipiger.sql
MMDL/SR/LCTL/BI-01-C PASS
MMDL/SR/LCTLS/BV-01-C PASS
MMDL/SR/LCTLS/BV-02-C PASS
MMDL/SR/LCTLS/BI-01-C FAIL https://hastebin.com/eguxuliweh.m
MMDL/SR/LCTLS/BI-02-C PASS
MMDL/SR/MLTEL/BV-02-C PASS

vikrant8051 commented:
@MariuszSkamra Ok.

vikrant8051 commented:
@MariuszSkamra @jhedberg 
MMDL/SR/GOO/BV-01-C FAIL https://hastebin.com/cequkamoju.sql
>> Failed ....because gen_onoff_get execute first instead of onoff_work_handler.
>> onoff_work_handler is defined in transaction.c & it get triggered by onoff_tt_handler.
>> onoff_tt_handler is timer function.

Same is happening in other cases. I will repair things shortly.

vikrant8051 commented:
@MariuszSkamra @jhedberg 
Now I think, most of the issues has resolved. 
There are four things which are still pending...I will work on it.
But meanwhile could you please re-run PTS for latest changes ?

vikrant8051 commented:
@MariuszSkamra 
Have you considered following commit (4th one) ?  
https://github.com/zephyrproject-rtos/zephyr/pull/8926/commits/47473edcf89805ce93f121f895c63a451b2f1b21

vikrant8051 commented:
@MariuszSkamra @jhedberg 
I have edit 3rd commit....so that now if received message has optional parameters then "GET" handler of that particular Sever will send acknowledgment which would contain Target value & TT.

vikrant8051 commented:
@MariuszSkamra Please re-run PTS against this modified PR.
Now only 3 issues are pending. I hope all others will get pass.

vikrant8051 commented:
@MariuszSkamra Kindly wait & don't run PTS ....... trying to implement Gen. Default TT Server & Client. 
I will inform you about status.

vikrant8051 commented:
@jhedberg 
I've added support for Generic Default tarnsition time Server/Client. 
But now I'm not able to get acknowledgment from any Server. Why this so ?
I need your help in this.

carlescufi commented:
recheck

carlescufi commented:
Close/reopen to fix CI

vikrant8051 commented:
@carlescufi what is CI ?

vikrant8051 commented:
@carlescufi Ok I got it.

carlescufi commented:
@vikrant8051 CI is continuous integration, the checks that run on Shippable

vikrant8051 commented:
@jhedberg 
Previous issue has solved .... 

/* Mapping of message handlers for Generic OnOff Client (0x1001) */
static const struct bt_mesh_model_op gen_onoff_cli_op[] = {
	{ BT_MESH_MODEL_OP_2(0x82, 0x04), **1**, gen_onoff_status },
	BT_MESH_MODEL_OP_END,
};

I had set 3 instead of 1... same for all other Servers.
But in this case why 1 ? 
because .... Acknowledgment could be = Present Value + Target Value + Remaining Time

vikrant8051 commented:
@MariuszSkamra @jhedberg 
Now only 2 issues are pending since I've also added "Generic Default transition Server/Client".
As per Mesh Model Specification, if Server received message with no #TT & #Delay (optional parameters) fields then it shall use values set by "Gen. Default Transition Time". Current implementation is also take care of it.

So if time allows you then please do review first & then execute PTS against latest implementation.

Thank You !!

vikrant8051 commented:
Hi @jhedberg @MariuszSkamra 

Waiting for review or new PTS logs. :)

Remaining two issues (regarding Range status) are confusing since don't know what exactly PTS
want from the Node.

vikrant8051 commented:
@jhedberg
Highlights of latest changes:

1) now value set by #gen_def_transition_time is going to impact other prime Servers

2) based on Transition Time field ( which could be default one or could be a part of message itself) tt_counter & tt_delta is calculated for each prime Servers. As per current implementation tt_counter could not be greater than 10.

3) Reason for previously failed PTS logs :
GET handler (which publishes status of node) executes before thread triggered by Server's timer. That thread was responsible to update present values. Because of that PTS dongle always received previously set value. 

4) Now if tt_counter = 0 then immediately that would update present values.

5) I've assumed that GET handler will always execute before timer's thread.

6) Any transaction in process will stop if receive new message.


I need your review/feedback regarding this implementations.

vikrant8051 commented:
@MariuszSkamra could you please re-run PTS for latest changes ?

vikrant8052 committed:
Merge branch 'master' into fix_bugs7
